// @telekom/scale-components/dist/collection/components/telekom/logo/logo.js
/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
import { Component, h, Prop, Host, Element } from '@stencil/core';
import classNames from 'classnames';
import statusNote from '../../../utils/status-note';
export class Logo {
  constructor() {
    /** (optional) Variant/color of the logo text and logo */
    this.variant = 'magenta';
    /** (optional) Set transparent background */
    this.transparent = false;
    /** (optional) Language of the logo text/ claimOff showes just the T Logo */
    this.language = 'en';
    /** (optional) The height in pixels */
    this.size = 36;
    /** (optional) When using the icon standalone, make it meaningful for accessibility */
    this.accessibilityTitle = 'Telekom Logo';
  }
  connectedCallback() {
    statusNote({ source: this.hostElement, tag: 'beta' });
  }
  styles() {
    return `:host {
      --logo-size: ${this.size}px;
    }`;
  }
  getLogoSvg() {
    return (h("scale-logo-svg", { language: this.language, color: this.variant, size: this.size, accessibilityTitle: this.accessibilityTitle }));
  }
  render() {
    return (h(Host, null,
      h("style", null, this.styles()),
      this.href === '' ? (h("div", { role: "img", "aria-label": "Telekom Logo", class: this.getCssClassMap(), onClick: this.clickHandler, tabIndex: 0 }, this.getLogoSvg())) : (h("a", { role: "link", "aria-label": "Telekom Logo", href: this.href, class: this.getCssClassMap(), onClick: this.clickHandler, tabIndex: 0 }, this.getLogoSvg()))));
  }
  getCssClassMap() {
    return classNames(`logo`, this.variant && `logo--variant-${this.variant}`, this.transparent && `logo--transparent`);
  }
  static get is() { return "scale-logo"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["./logo.css"]
  }; }
  static get styleUrls() { return {
    "$": ["logo.css"]
  }; }
  static get properties() { return {
    "variant": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "'magenta' | 'white'",
        "resolved": "\"magenta\" | \"white\"",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "(optional) Variant/color of the logo text and logo"
      },
      "attribute": "variant",
      "reflect": false,
      "defaultValue": "'magenta'"
    },
    "transparent": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "(optional) Set transparent background"
      },
      "attribute": "transparent",
      "reflect": false,
      "defaultValue": "false"
    },
    "language": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "| 'de'\n    | 'en'\n    | 'cz'\n    | 'hr'\n    | 'hu'\n    | 'me'\n    | 'mk_lat'\n    | 'mk_kyr'\n    | 'ro'\n    | 'sk'\n    | ''",
        "resolved": "\"\" | \"cz\" | \"de\" | \"en\" | \"hr\" | \"hu\" | \"me\" | \"mk_kyr\" | \"mk_lat\" | \"ro\" | \"sk\"",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "(optional) Language of the logo text/ claimOff showes just the T Logo"
      },
      "attribute": "language",
      "reflect": false,
      "defaultValue": "'en'"
    },
    "size": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "(optional) The height in pixels"
      },
      "attribute": "size",
      "reflect": false,
      "defaultValue": "36"
    },
    "href": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "(optional) Set a link"
      },
      "attribute": "href",
      "reflect": false
    },
    "clickHandler": {
      "type": "any",
      "mutable": false,
      "complexType": {
        "original": "any",
        "resolved": "any",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "(optional) Possibility for adding a onClick Event"
      },
      "attribute": "click-handler",
      "reflect": false
    },
    "accessibilityTitle": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "(optional) When using the icon standalone, make it meaningful for accessibility"
      },
      "attribute": "accessibility-title",
      "reflect": false,
      "defaultValue": "'Telekom Logo'"
    }
  }; }
  static get elementRef() { return "hostElement"; }
}
