// @telekom/scale-components/dist/esm/scale-tab-header.entry.js
import { r as registerInstance, h, H as Host, g as getElement } from './index-5fb2d2c0.js';
import { c as classnames } from './index-349fc524.js';

const tabHeaderCss = ":host{outline:none;--font-family:var(--scl-font-family-sans, TeleNeoWeb);--font-weight:var(--scl-font-weight-bold);--height-small:2rem;--color-hover:var(--scl-color-primary-hover);--color-selected:var(--scl-color-primary);--background-underline-hover:var(--scl-color-primary-hover);--background-underline-active:var(--scl-color-primary-active);--color-active:var(--scl-color-primary-active);--box-shadow-focus:0 0 0 var(--scl-spacing-2) var(--scl-color-focus);--spacing-right-slotted:var(--scl-spacing-8)}.tab-header{cursor:pointer;display:inline-flex;outline:none;position:relative;box-sizing:border-box;align-items:center;padding-left:0;height:var(--tab-height);margin-right:var(--tab-spacing);padding-right:var(--tab-spacing);font-family:var(--font-family);font-size:calc(var(--tab-height) / 2);font-weight:var(--scl-font-weight-bold);border-radius:var(--tab-radius)}.tab-header--small{--tab-height:var(--height-small)}.tab-header:after{left:0;right:0;width:100%;bottom:0;content:'';display:block;position:absolute;height:var(--tab-border-size);background-color:var(--tab-border-color)}.tab-header:before{left:0;right:0;width:100%;bottom:0;height:0;content:'';display:block;z-index:1;position:absolute;border-top:1px solid transparent}.tab-header:hover{color:var(--color-hover)}.tab-header:hover:after{background-color:var(--background-underline-hover)}.tab-header:active{color:var(--color-active)}.tab-header:active:after{background-color:var(--background-underline-active)}.tab-header ::slotted(*){margin-right:var(--spacing-right-slotted)}.tab-header--has-focus{box-shadow:var(--box-shadow-focus)}.tab-header--selected{color:var(--color-selected)}.tab-header--selected:after{border-top-left-radius:var(--scl-radius-8);border-top-right-radius:var(--scl-radius-8);height:var(--tab-border-size-selected);background-color:var(--color-selected)}.tab-header--selected:before{border-top:var(--tab-border-size-selected) solid transparent}";

let i = 0;
const TabHeader = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.generatedId = i++;
    /** True for smaller height and font size */
    this.small = false;
    this.hasFocus = false;
  }
  selectedChanged(newValue) {
    if (newValue === true) {
      // Having focus on the host element, and not on inner elements,
      // is required because screen readers.
      this.el.focus();
    }
    this.updateSlottedIcon();
  }
  /**
   * Find slotted icons, and if any, add the `selected` attribute accordingly.
   */
  updateSlottedIcon() {
    const slot = this.container.querySelector('slot');
    if (slot === null) {
      return;
    }
    const children = Array.from(slot.assignedNodes())
      .filter((node) => node.nodeType === 1)
      .filter((node) => node.nodeName.toUpperCase().indexOf('ICON') > -1);
    if (children.length === 0) {
      return;
    }
    const action = this.selected ? 'setAttribute' : 'removeAttribute';
    children.forEach((child) => child[action]('selected', ''));
  }
  render() {
    return (h(Host, { id: `scale-tab-header-${this.generatedId}`, role: "tab", "aria-selected": String(this.selected), tabindex: this.selected ? '0' : '-1', onFocus: () => (this.hasFocus = true), onBlur: () => (this.hasFocus = false) }, this.styles && h("style", null, this.styles), h("span", { part: this.getBasePartMap(), ref: (el) => (this.container = el), class: this.getCssClassMap() }, h("slot", null))));
  }
  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }
  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }
  getCssOrBasePartMap(mode) {
    const component = 'tab-header';
    const prefix = mode === 'basePart' ? '' : `${component}--`;
    return classnames(component, this.selected && `${prefix}selected`, this.small && `${prefix}small`, this.hasFocus && `${prefix}has-focus`);
  }
  get el() { return getElement(this); }
  static get watchers() { return {
    "selected": ["selectedChanged"]
  }; }
};
TabHeader.style = tabHeaderCss;

export { TabHeader as scale_tab_header };
