// @telekom/scale-components/dist/esm/scale-tab-panel.entry.js
import { r as registerInstance, h, H as Host } from './index-5fb2d2c0.js';
import { c as classnames } from './index-349fc524.js';

const tabPanelCss = ":host{--spacing-top:var(--scl-spacing-24);--spacing-top-small:var(--scl-spacing-16)}.tab-panel{padding-top:var(--spacing-top)}.tab-panel--small{padding-top:var(--spacing-top-small)}";

let i = 0;
const TabPanel = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.generatedId = i++;
    /** True for smaller height and font size */
    this.small = false;
  }
  render() {
    return (h(Host, { id: `scale-tab-panel-${this.generatedId}`, role: "tabpanel" }, this.styles && h("style", null, this.styles), h("div", { part: this.getBasePartMap(), class: this.getCssClassMap() }, h("slot", null))));
  }
  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }
  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }
  getCssOrBasePartMap(mode) {
    const component = 'tab-panel';
    const prefix = mode === 'basePart' ? '' : `${component}--`;
    return classnames(component, this.small && `${prefix}small`);
  }
};
TabPanel.style = tabPanelCss;

export { TabPanel as scale_tab_panel };
