// @telekom/scale-components/dist/esm/scale-logo.entry.js
import { r as registerInstance, h, H as Host, g as getElement } from './index-5fb2d2c0.js';
import { c as classnames } from './index-349fc524.js';
import { s as statusNote } from './status-note-d227f142.js';

const logoCss = ":host{--logo-size:36px}.logo{display:inline-flex;height:var(--logo-size);padding:calc(var(--logo-size) / 3);position:relative}.logo--variant-magenta{background-color:var(--scl-color-white)}.logo--variant-white{background-color:var(--scl-color-primary)}.logo--transparent{background-color:transparent}";

const Logo = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    /** (optional) Variant/color of the logo text and logo */
    this.variant = 'magenta';
    /** (optional) Set transparent background */
    this.transparent = false;
    /** (optional) Language of the logo text/ claimOff showes just the T Logo */
    this.language = 'en';
    /** (optional) The height in pixels */
    this.size = 36;
    /** (optional) When using the icon standalone, make it meaningful for accessibility */
    this.accessibilityTitle = 'Telekom Logo';
  }
  connectedCallback() {
    statusNote({ source: this.hostElement, tag: 'beta' });
  }
  styles() {
    return `:host {
      --logo-size: ${this.size}px;
    }`;
  }
  getLogoSvg() {
    return (h("scale-logo-svg", { language: this.language, color: this.variant, size: this.size, accessibilityTitle: this.accessibilityTitle }));
  }
  render() {
    return (h(Host, null, h("style", null, this.styles()), this.href === '' ? (h("div", { role: "img", "aria-label": "Telekom Logo", class: this.getCssClassMap(), onClick: this.clickHandler, tabIndex: 0 }, this.getLogoSvg())) : (h("a", { role: "link", "aria-label": "Telekom Logo", href: this.href, class: this.getCssClassMap(), onClick: this.clickHandler, tabIndex: 0 }, this.getLogoSvg()))));
  }
  getCssClassMap() {
    return classnames(`logo`, this.variant && `logo--variant-${this.variant}`, this.transparent && `logo--transparent`);
  }
  get hostElement() { return getElement(this); }
};
Logo.style = logoCss;

export { Logo as scale_logo };
