// @telekom/scale-components/dist/esm/scale-app-shell.entry.js
import { r as registerInstance, h, H as Host, g as getElement } from './index-5fb2d2c0.js';
import { c as classnames } from './index-349fc524.js';

const appShellCss = ":host{--background:var(--scl-color-background-standard, #fff);--spacing-x:var(--scl-spacing-24);--radius:var(--scl-radius-8);--header-nav-height:56px;--header-brand-height:72px;--header-border-radius:var(--radius);--header-transition-speed:0.3s;--header-brand-collapsed-height:4px;--header-max-width:inherit}.sr-only{position:absolute;left:-10000px;overflow:hidden}.content{background:var(--background);padding-left:var(--spacing-x);padding-right:var(--spacing-x)}@media (max-width: 1023px){.shell .content{padding-top:calc(\n      var(--header-brand-collapsed-height) + var(--header-nav-height)\n    )}}@media (min-width: 1024px){.shell.shell--sticky .content{padding-top:calc(\n      var(--header-brand-collapsed-height) + var(--header-nav-height)\n    )}.content{transition:padding-top var(--header-transition-speed) ease-in-out;padding-top:calc(var(--header-brand-height) + var(--header-nav-height))}}@media (min-width: 1552px){.shell .content{margin:0 auto;max-width:var(--header-max-width)}}";

const Shell = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.portalName = '';
    this.claimLang = 'de';
    this.mainNavigation = [];
    this.iconNavigation = [];
    this.sectorNavigation = [];
    this.addonNavigation = [];
    this.activeRouteId = '';
    this.activeSectorId = '';
    this.scrolled = false;
  }
  onScroll() {
    this.scrolled = window.pageYOffset > 2;
  }
  componentWillLoad() {
    this.hasSlotHeader = !!this.hostElement.querySelector('[slot="header"]');
  }
  componentWillUpdate() { }
  disconnectedCallback() { }
  render() {
    return (h(Host, null, this.styles && h("style", null, this.styles), h("div", { part: classnames('base', this.scrolled && 'sticky'), class: this.getCssClassMap() }, this.hasSlotHeader ? (h("slot", { name: "header" })) : (h("scale-app-header", { logoClick: this.logoClick, logoHref: this.logoHref, logoTitle: this.logoTitle, portalName: this.portalName, mainNavigation: this.mainNavigation, iconNavigation: this.iconNavigation, sectorNavigation: this.sectorNavigation, addonNavigation: this.addonNavigation, activeRouteId: this.activeRouteId, activeSectorId: this.activeSectorId, claimLang: this.claimLang })), h("main", { class: "content" }, h("slot", null)))));
  }
  getCssClassMap() {
    return classnames('shell', this.scrolled && 'shell--sticky');
  }
  get hostElement() { return getElement(this); }
};
Shell.style = appShellCss;

export { Shell as scale_app_shell };
