// @telekom/scale-components/dist/cjs/scale-tab-panel.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-66a8d2d7.js');
const index$1 = require('./index-550fcacd.js');

const tabPanelCss = ":host{--spacing-top:var(--scl-spacing-24);--spacing-top-small:var(--scl-spacing-16)}.tab-panel{padding-top:var(--spacing-top)}.tab-panel--small{padding-top:var(--spacing-top-small)}";

let i = 0;
const TabPanel = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.generatedId = i++;
    /** True for smaller height and font size */
    this.small = false;
  }
  render() {
    return (index.h(index.Host, { id: `scale-tab-panel-${this.generatedId}`, role: "tabpanel" }, this.styles && index.h("style", null, this.styles), index.h("div", { part: this.getBasePartMap(), class: this.getCssClassMap() }, index.h("slot", null))));
  }
  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }
  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }
  getCssOrBasePartMap(mode) {
    const component = 'tab-panel';
    const prefix = mode === 'basePart' ? '' : `${component}--`;
    return index$1.classnames(component, this.small && `${prefix}small`);
  }
};
TabPanel.style = tabPanelCss;

exports.scale_tab_panel = TabPanel;
