// @telekom/scale-components/dist/cjs/scale-app-shell.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-66a8d2d7.js');
const index$1 = require('./index-550fcacd.js');

const appShellCss = ":host{--background:var(--scl-color-background-standard, #fff);--spacing-x:var(--scl-spacing-24);--radius:var(--scl-radius-8);--header-nav-height:56px;--header-brand-height:72px;--header-border-radius:var(--radius);--header-transition-speed:0.3s;--header-brand-collapsed-height:4px;--header-max-width:inherit}.sr-only{position:absolute;left:-10000px;overflow:hidden}.content{background:var(--background);padding-left:var(--spacing-x);padding-right:var(--spacing-x)}@media (max-width: 1023px){.shell .content{padding-top:calc(\n      var(--header-brand-collapsed-height) + var(--header-nav-height)\n    )}}@media (min-width: 1024px){.shell.shell--sticky .content{padding-top:calc(\n      var(--header-brand-collapsed-height) + var(--header-nav-height)\n    )}.content{transition:padding-top var(--header-transition-speed) ease-in-out;padding-top:calc(var(--header-brand-height) + var(--header-nav-height))}}@media (min-width: 1552px){.shell .content{margin:0 auto;max-width:var(--header-max-width)}}";

const Shell = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.portalName = '';
    this.claimLang = 'de';
    this.mainNavigation = [];
    this.iconNavigation = [];
    this.sectorNavigation = [];
    this.addonNavigation = [];
    this.activeRouteId = '';
    this.activeSectorId = '';
    this.scrolled = false;
  }
  onScroll() {
    this.scrolled = window.pageYOffset > 2;
  }
  componentWillLoad() {
    this.hasSlotHeader = !!this.hostElement.querySelector('[slot="header"]');
  }
  componentWillUpdate() { }
  disconnectedCallback() { }
  render() {
    return (index.h(index.Host, null, this.styles && index.h("style", null, this.styles), index.h("div", { part: index$1.classnames('base', this.scrolled && 'sticky'), class: this.getCssClassMap() }, this.hasSlotHeader ? (index.h("slot", { name: "header" })) : (index.h("scale-app-header", { logoClick: this.logoClick, logoHref: this.logoHref, logoTitle: this.logoTitle, portalName: this.portalName, mainNavigation: this.mainNavigation, iconNavigation: this.iconNavigation, sectorNavigation: this.sectorNavigation, addonNavigation: this.addonNavigation, activeRouteId: this.activeRouteId, activeSectorId: this.activeSectorId, claimLang: this.claimLang })), index.h("main", { class: "content" }, index.h("slot", null)))));
  }
  getCssClassMap() {
    return index$1.classnames('shell', this.scrolled && 'shell--sticky');
  }
  get hostElement() { return index.getElement(this); }
};
Shell.style = appShellCss;

exports.scale_app_shell = Shell;
