// @telekom/scale-components/dist/collection/components/loading-spinner/loading-spinner.js
import { Component, h, Host, Prop } from '@stencil/core';
import classNames from 'classnames';
export class LoadingSpinner {
  constructor() {
    this.variant = 'primary';
    this.alignment = 'horizontal';
    this.size = 'small';
  }
  render() {
    return (h(Host, null,
      h("div", { part: this.getBasePartMap(), class: this.getCssClassMap() },
        h("div", { part: "container", class: "spinner__container" },
          h("div", { class: "spinner__circle-background" }),
          h("div", { part: "circle", class: "spinner__circle" },
            h("div", null),
            h("div", null),
            h("div", null),
            h("div", null))),
        h("div", { class: "sr-only", "aria-live": "polite" }, this.text || 'Loading'),
        h("div", { part: "text", class: "spinner__text", "aria-hidden": "true" }, this.text))));
  }
  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }
  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }
  getCssOrBasePartMap(mode) {
    const name = 'spinner';
    const prefix = mode === 'basePart' ? '' : `${name}--`;
    return classNames(name, this.alignment && `${prefix}alignment-${this.alignment}`, this.variant && `${prefix}variant-${this.variant}`, this.size && `${prefix}size-${this.size}`, this.text && `${prefix}text`);
  }
  static get is() { return "scale-loading-spinner"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["./loading-spinner.css"]
  }; }
  static get styleUrls() { return {
    "$": ["loading-spinner.css"]
  }; }
  static get properties() { return {
    "variant": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "'white' | 'primary'",
        "resolved": "\"primary\" | \"white\"",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "variant",
      "reflect": false,
      "defaultValue": "'primary'"
    },
    "alignment": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "'horizontal' | 'vertical'",
        "resolved": "\"horizontal\" | \"vertical\"",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "alignment",
      "reflect": false,
      "defaultValue": "'horizontal'"
    },
    "text": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "text",
      "reflect": false
    },
    "size": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "'small' | 'large'",
        "resolved": "\"large\" | \"small\"",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "size",
      "reflect": false,
      "defaultValue": "'small'"
    }
  }; }
}
